### PR TITLE
AIP-6

### PR DESCRIPTION
**Pull request by nithinkrishna**
Changes to be released described in [#AIP6](https://aips.ampleforth.org/AIPs/aip-6)

### CODE REVIEW

File: contracts/Orchestrator.sol [hunk 1]
```diff
@@ -123,12 +101,10 @@ contract Orchestrator is Ownable {
      * @param data The encoded data payload.
      * @return True on success
      */
-    function externalCall(address destination, bytes data)
-        internal
-        returns (bool)
-    {
+    function externalCall(address destination, bytes memory data) internal returns (bool) {
```
Comment (thegostep) [inline contracts/Orchestrator.sol:104]: should probably replace the assembly with a .call

File: contracts/UFragments.sol [hunk 2]
```diff
@@ -67,23 +66,35 @@ contract UFragments is ERC20Detailed, Ownable {
     uint256 private constant TOTAL_GONS = MAX_UINT256 - (MAX_UINT256 % INITIAL_FRAGMENTS_SUPPLY);
 
     // MAX_SUPPLY = maximum integer < (sqrt(4*TOTAL_GONS + 1) - 1) / 2
-    uint256 private constant MAX_SUPPLY = ~uint128(0);  // (2^128) - 1
+    uint256 private constant MAX_SUPPLY = ~uint128(0); // (2^128) - 1
```
Comment (thegostep) [inline contracts/UFragments.sol:69]: `type(uint128).max`

File: contracts/UFragments.sol [hunk 3]
```diff
@@ -98,42 +109,43 @@ contract UFragments is ERC20Detailed, Ownable {
         onlyMonetaryPolicy
         returns (uint256)
     {
+        uint256 oldTotalSupply = _totalSupply;
```
Comment (thegostep) [inline contracts/UFragments.sol:112]: consider letting the compiler do this optimization

Comment (thegostep) [reply contracts/UFragments.sol:112]: by using new version + optimizer

File: contracts/UFragments.sol [hunk 4]
```diff
@@ -273,17 +356,53 @@ contract UFragments is ERC20Detailed, Ownable {
...
+    function decreaseAllowance(address spender, uint256 subtractedValue) external returns (bool) {
         uint256 oldValue = _allowedFragments[msg.sender][spender];
+        uint256 newValue;
         if (subtractedValue >= oldValue) {
-            _allowedFragments[msg.sender][spender] = 0;
+            newValue = 0;
         } else {
-            _allowedFragments[msg.sender][spender] = oldValue.sub(subtractedValue);
+            newValue = oldValue.sub(subtractedValue);
         }
```
Comment (thegostep) [inline contracts/UFragments.sol:366]: `uint256 newValue = subtractedValue >= oldValue ? 0 : oldValue.sub(subtractedValue);`

**Review by thegostep — COMMENTED** [review]
some suggestions

File: contracts/Orchestrator.sol (hunk 1 above)
Comment (nithinkrishna) [reply contracts/Orchestrator.sol:104]: Makes sense 👍  We can address the Orchestrator changes separately ..
https://github.com/ampleforth/uFragments/issues/192

File: contracts/UFragments.sol (hunk 2 above)
Comment (nithinkrishna) [reply contracts/UFragments.sol:69]: https://github.com/ampleforth/uFragments/pull/196

File: contracts/UFragments.sol (hunk 3 above)
Comment (nithinkrishna) [reply contracts/UFragments.sol:112]: https://github.com/ampleforth/uFragments/pull/197

File: contracts/UFragments.sol (hunk 4 above)
Comment (nithinkrishna) [reply contracts/UFragments.sol:366]: https://github.com/ampleforth/uFragments/pull/198

File: contracts/UFragmentsPolicy.sol [hunk 5]
```diff
@@ -80,7 +83,7 @@ contract UFragmentsPolicy is Ownable {
     // Both are 18 decimals fixed point numbers.
     uint256 private constant MAX_RATE = 10**6 * 10**DECIMALS;
     // MAX_SUPPLY = MAX_INT256 / MAX_RATE
-    uint256 private constant MAX_SUPPLY = ~(uint256(1) << 255) / MAX_RATE;
+    uint256 public constant MAX_SUPPLY = uint256(type(int256).max) / MAX_RATE;
```
Comment (nithinkrishna) [inline contracts/UFragmentsPolicy.sol:86]: @nithinkrishna Turn this back to private

**Review by brandoniles — APPROVED** [review]
LGTM

**Pull request closed without merging**